scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryITUtils.scala

```diff
@@ -0,0 +1,149 @@
+
+  private val mapping = MMap.empty[TableReference, TableReference]
+
+  /**
+   * Mock a BigQuery table. Each table can be mocked only once in a test class.
+   */
+  def mockTable(original: String): MockTable = mockTable(BigQueryIO.parseTableSpec(original))
+
+  /**
+   * Mock a BigQuery table. Each table can be mocked only once in a test class.
+   */
+  def mockTable(original: TableReference): MockTable = {
+    require(
+      !mapping.contains(original),
+      s"Table ${BigQueryIO.toTableSpec(original)} already registered for mocking")
+
+    val t = bq.getTable(original)
+    val temp = bq.temporaryTable(t.getLocation)
+    mapping += (original -> temp)
+    new MockTable(bq, t.getSchema, original, temp)
+  }
+
+  /**
+   * Perform a live query against BigQuery service, substituting mocked tables with test data.
+   */
+  def query(sqlQuery: String, flattenResults: Boolean = false): Seq[TableRow] = {
+    val isLegacy = bq.isLegacySql(sqlQuery, flattenResults)
+    val mockQuery = mapping.foldLeft(sqlQuery){ case (q, (src, dst)) =>
+      q.replace(toTableSpec(src, isLegacy), toTableSpec(dst, isLegacy))
+    }
+    bq.getQueryRows(mockQuery, flattenResults).toList
+  }
+
+  /**
+   * Perform a live query against BigQuery service, substituting mocked tables with test data.
+   */
+  def typedQuery[T <: HasAnnotation : ClassTag : TypeTag]
+  (sqlQuery: String, flattenResults: Boolean = false): Seq[T] = {
+    val bqt = BigQueryType[T]
+    query(sqlQuery, flattenResults).map(bqt.fromTableRow)
+  }
+
+  private def toTableSpec(table: TableReference, isLegacy: Boolean) =
+    if (isLegacy) {
+      s"[${table.getProjectId}:${table.getDatasetId}.${table.getTableId}]"
+    } else {
+      s"`${table.getProjectId}.${table.getDatasetId}.${table.getTableId}`"
+    }
+
+}
+
+/**
+ * A BigQuery table being mocked for test.
+ */
+class MockTable(private val bq: BigQueryClient,
+                private val schema: TableSchema,
+                private val original: TableReference,
+                private val temp: TableReference) {
+
+  private var mocked: Boolean = false
+
+  private def ensureUnique(): Unit = {
+    require(!mocked, s"Table ${BigQueryIO.toTableSpec(original)} already populated with mock data")
+    this.mocked = true
+  }
+
+  private def writeRows(rows: Seq[TableRow]): Unit = {
+    bq.writeTableRows(temp, rows.toList, schema, WRITE_EMPTY, CREATE_IF_NEEDED)
+  }
+
+  /**
+   * Populate the table with mock data.
+   */
+  def withData(rows: Seq[TableRow]): Unit = {
+    ensureUnique()
+    writeRows(rows)
+  }
+
+  /**
+   * Populate the table with mock data.
+   */
+  def withTypedData[T <: HasAnnotation : ClassTag : TypeTag](rows: Seq[T]): Unit = {
+    val bqt = BigQueryType[T]
+    withData(rows.map(bqt.toTableRow))
+  }
+
+  /**
+   * Populate the table with sample data from the original table.
+   * Note that rows are taken from the beginning of the table and not truly random.
```

Comment:
I can imagine this being a problem with a lot of tables - could we use reservoir sampling? Or should we wait until someone asks for it :)